Flickr

Akiat opened the pull request:
The flickr driver is normally finished. Some flake8 errors are present but it is to improve the code lisibility. If needed I can fix that.

Akiat committed:
Flickr basics: upload, download, move, remove, replace

Akiat committed:
Add flickr auth script and provide api client/secret

Akiat committed:
clean (one way driver)

Khady committed:
Fix flake8

Khady commented:
Are the `global flickr` required in every function?

Akiat commented:
Not in the handlers, we can remove them.

Khady committed:
Remove the 'global flickr' in handlers

Khady commented:
Seems fine. :+1:

k4nar commented:
I'll review it in a shot time.

k4nar commented:
`get_tokens` is not Python 3 compatible, but we won't need this script when #97 will land anyway, so that's not a big deal.

k4nar commented on onitu/drivers/flickr/flickr.py:
```diff
@@ -0,0 +1,335 @@
+import cStringIO
```
Not compatible with Python 3.

k4nar commented on onitu/drivers/flickr/flickr.py:
```diff
@@ -0,0 +1,335 @@
+        })
+
+        if photo_id:
+            print 'replace'
```
Please remove debugging

k4nar commented on onitu/drivers/flickr/flickr.py:
```diff
@@ -0,0 +1,335 @@
+        prepared = dummy_req.prepare()
+        headers = prepared.headers
+
+        fileobj = cStringIO.StringIO(content)
```
Warning: In Python 3 you must use `io.StringIO` or `io.ByteIO` regarding the type of your data.

k4nar commented on onitu/drivers/flickr/flickr.py:
```diff
@@ -0,0 +1,335 @@
+        return self.search_file(metadata)
+
+    def check_error(self, func_name, req):
+        if req.status_code != 200:
```
Something like `req.status_code // 100 != 2` would be better I think (in order to accept 201, 203, etc…)

k4nar commented on onitu/drivers/flickr/flickr.py:
```diff
@@ -0,0 +1,335 @@
+
+    def call(self, method, url, params, content=None):
+        # Clean params
+        for k in params.keys():
```
I don't think that's a good idea to remove keys when iterating over a dict. This would be better :

``` python
params = {k: v for k, v in params.items() if v is not None}
```

k4nar commented on onitu/drivers/flickr/flickr.py:
```diff
@@ -0,0 +1,335 @@
+        return self.onitu_tag + self.root + '/' + hashlib.md5(
+            metadata.filename + str(metadata.size)).hexdigest()
+
+    def load_base_params(self):
```
This function is really dangerous ! You shouldn't assign a property, but only return an array. This way, you wouldn't have to clean the dict after having called "update". Every call should return a new dict, which should be explicitly given to the functions needing it.

k4nar commented on onitu/drivers/flickr/flickr.py:
```diff
@@ -0,0 +1,335 @@
+            r = requests.post(url, params=params, auth=self.oauth)
+
+        # clean base_params
+        self.load_base_params()
```
This is why `load_base_params` is such a bad function :) .

k4nar commented on onitu/drivers/flickr/flickr.py:
```diff
@@ -0,0 +1,335 @@
+        else:
+            return json['photos']['photo'][0]['id']
+
+    def call(self, method, url, params, content=None):
```
I'm not a big fan of this function. You should have two methods instead, one to get and one to upload.

k4nar commented on onitu/drivers/flickr/flickr.py:
```diff
@@ -0,0 +1,335 @@
+        metadata.write()
+
+    def get_file(self, metadata):
+        """return nothing"""
```
I'm pretty sure that "get_file" should return something :D . Plus, you have `return r.content` just after.

k4nar commented on onitu/drivers/flickr/flickr.py:
```diff
@@ -0,0 +1,335 @@
+        if req.status_code != 200:
+            raise ServiceError(
+                '{}: Status code {} received'.format(
+                    func_name, req.status_code
```
Doesn't Flickr return a body with a message explaining the error ?

Akiat replied:
Nop, the body contains only the full url used for the call.

k4nar commented on onitu/drivers/flickr/flickr.py:
```diff
@@ -0,0 +1,335 @@
+            return r.content
+
+    def delete_file(self, metadata):
+        """return nothing"""
```
Not relevant.

k4nar commented on onitu/drivers/flickr/flickr.py:
```diff
@@ -0,0 +1,335 @@
+        try:
+            tag = self.get_file_info(photo_id)['photo']['tags']['tag']
+        except KeyError:
+            global plug
```
You don't need that.

k4nar commented on onitu/drivers/flickr/flickr.py:
```diff
@@ -0,0 +1,335 @@
+                "get_tag_id: Cannot find tag id of tag '{}'".format(tag_name))
+            return None
+        i = 0
+        while (tag[i]):
```
Why not `for element in tag` ? This is really not Pythonic.

k4nar commented on onitu/drivers/flickr/flickr.py:
```diff
@@ -0,0 +1,335 @@
+    def move_file(self, old, new):
+        photo_id = self.load_photo_id(old)
+        new_tag = self.create_tag(new)
+        if photo_id:
```
Why `photo_id` could be None ? And if so, shouldn't we print an error ? Same thing for `tag_id`.

k4nar commented on onitu/drivers/flickr/flickr.py:
```diff
@@ -0,0 +1,335 @@
+
+
+def start():
+    global plug
```
Not needed.

k4nar commented on onitu/drivers/flickr/flickr.py:
```diff
@@ -0,0 +1,335 @@
+            return metadata.extra['id']
+        return self.search_file(metadata)
+
+    def check_error(self, func_name, req):
```
I don't think that we should have this function. Instead, we should have a `self.request` where we make the request inside a try/except.

Akiat replied:
No exception is raised. So I must check the status_code, it's just to avoid to check it each time.

k4nar replied:
You can have an exception if Flickr is down for instance. And why wouldn't you check it each time ? The check is very quick.

k4nar commented on onitu/drivers/flickr/manifest.json:
```diff
@@ -0,0 +1,25 @@
+{
+  "name": "Flickr",
+  "description": "Driver to synchronize files with a Flickr account.",
```
Maybe we could come up with a better description. Like what kind of file are supported, what are the limitations, etc…
Of course the description should be short, but it should also be instructive.

Akiat committed:
Fix pull request issues

k4nar commented:
We need to rebase it to master and adapt it to the new architecture. Cf 73b2b8ea27359ada789514908ac7900ee5f28aae to see how to do it.
We also need to squash the commit into one.

@Akiat can you do it this week ?

k4nar commented on onitu/drivers/flickr/flickr.py:
```diff
@@ -0,0 +1,346 @@
+# ############################## UTILS ######################################
+
+    def create_tag(self, metadata):
+        return self.onitu_tag + self.root + '/' + hashlib.md5(
```
Could you rewrite it using `format` instead of `+` ? This is more efficient and understandable.

k4nar commented on onitu/drivers/flickr/flickr.py:
```diff
@@ -0,0 +1,346 @@
+            metadata.filename + str(metadata.size)).hexdigest()
+
+    def load_tag_id(self, metadata):
+        if 'tag_id' in metadata.extra:
```
As far as I can see you never actually put it in `metadata.extra`.

k4nar commented on onitu/drivers/flickr/flickr.py:
```diff
@@ -0,0 +1,346 @@
+            raise ServiceError(
+                'Call: Status code {} received.'.format(
+                    r.status_code
+                    )
```
I'm surprised that pyflakes lets you do that. Anyway, you should move that bracket and the one bellow one tab left.

k4nar commented on onitu/drivers/flickr/flickr.py:
```diff
@@ -0,0 +1,346 @@
+                'Content-Type': m.content_type}
+
+        try:
+            r = requests.post(url, data=m, headers=auth)
```
You should use your `self.call` method.

k4nar commented on onitu/drivers/flickr/flickr.py:
```diff
@@ -0,0 +1,346 @@
+            return metadata.extra['id']
+        return self.search_file(metadata)
+
+    def call(self, func, url, params=None):
```
It should accept a `**kwargs` argument and give it to `func`.

k4nar commented on onitu/drivers/flickr/flickr.py:
```diff
@@ -0,0 +1,346 @@
+
+        self.call(requests.post, self.rest_url, params)
+
+# ############################# PHOTOSETS ###################################
```
Do we need those two functions ?

Akiat replied:
Yes for the move

k4nar replied:
I was talking about `list_photosets` and `create_photoset`

k4nar commented on onitu/drivers/flickr/flickr.py:
```diff
@@ -0,0 +1,346 @@
+        self.rest_url = 'https://api.flickr.com/services/rest/'
+        self.upload_url = 'https://up.flickr.com/services/upload/'
+        self.replace_url = 'https://up.flickr.com/services/replace/'
+        self.onitu_tag = 'ONITU_TAG_DO_NOT_REMOVE/'
```
Could you add a small explanation on why that's needed ? Also if the users are able to see it, i would rather rename it to only "Onitu", and explain in the Manifest that they should not remove this tag.

Akiat committed:
new architecture and rebase

Akiat committed:
delete old driver files

Akiat committed:
PR issues fixed

k4nar commented on drivers/flickr/onitu_flickr/flickr.py:
```diff
@@ -0,0 +1,348 @@
+            return metadata.extra['id']
+        return self.search_file(metadata)
+
+    def call(self, func, url, kwargs):
```
It should take a `**kwargs`, it would make your call cleaner.

For example : 

```
kwargs = dict(params=params, auth=self.oauth)
r = self.call(requests.get, self.rest_url, kwargs)
```

Could be expressed as : 

```
r = self.call(requests.get, self.rest_url, params=params, auth=self.oauth)
```

k4nar replied:
Actually you could define it as `def call(self, func, url, params=None, auth=None, **kwargs):`, it would be even better.
Also, if you always give the oauth token, add it in this function directly.

k4nar commented on drivers/flickr/onitu_flickr/flickr.py:
```diff
@@ -0,0 +1,348 @@
+        return file id otherwise """
+        tag = self.create_tag(metadata)
+
+        params = {
```
I really think that your code would be far simpler with a `get_params` function, accepting some keyword arguments.

Akiat committed:
Add client id/secret directly in the manifest

Akiat committed:
last modif for PR

Akiat committed:
Add the root album

Akiat committed:
fix pep8

Scylardor committed:
Flickr rework first pass: Use flickrapi library

Use flickrapi to do all the heavy lifting of authentication/Flickr
API calling for us. Much of Onitu code will be reworked, or made
useless, by this change.

- Rework the get_tokens.py script in order to work with flickrapi

Scylardor committed:
Update Flickr: check changes thread is OK

- Handlers still work in progress.

Scylardor committed:
Fix the "handlers don't get called" issue. Bidirectional transfer OK

The problem was we didn't set metadata.size on new photo detection
on Flickr. The bigger problem behind that is that Flickr gives us no
way to know the real size in bytes of its files. Our somewhat dirty
solution to this is to set a TEMPORARY metadata.size of 1 to trigger
get_file if other drivers are interested, and then update the size
in get_file, with the length of the data chunk.

- get_file, upload_file, delete_file handlers are OK

- Add a variety of functions in FlickrHandler class to easily manage
handlers needs, particularly the upload_file which has a lot
of caveats to avoid. To list a few:
   o check if the photo already exist or not
   o create the photoset if it doesn't exist
   o (even trickier) Recreate the photoset if it doesn't exist ANYMORE
     (thus Onitu having an obsolete photoset ID)

- Modify the check changes thread to handle the obsolete photoset IDs

Scylardor committed:
Flickr driver move file handler OK

- Update 'last_update' timestamp in upload_file and move_file to
avoid an unnecessary transfer in the change checking thread

Scylardor committed:
Add of a PEP8 clean version of Flickrapi.

Onitu now acts as a vendor for Flickrapi. Minor changes have been
done to be PEP8 compliant (damn you, nazi flake8 !)

Scylardor committed:
Flickr driver: Make the necessary changes to use flickrapi

Flickrapi isn't a global package anymore so we have to change imports.

- First version of the Flickr test suite, trying to find a way to
make them pass by fooling Flickr's image detection system (we don't
generate images for tests).

- Update driver requirements (added the `six' package)

Scylardor committed:
Fix PEP8 overcleaning in flickrapi

Scylardor committed:
Flickr driver: Fix imports to make it work in Python 3

Scylardor commented:
Bump.

I reworked the driver by using the Flickrapi library. It does most of the heavy lifting of OAuth authentication and HTTP requests handling for us.

Unlike the driver's first version it is now bidirectional. We can transfer photos both from and to a given Flickr photoset.

I'm currently facing a testing issue, since to test with Flickr we need to send images, not random garbage data. I made some modifications in the code and the manifest to be able to activate a "PNG header" mode that would prepend all data sent to the driver with a PNG header, to make the data look like an image to Flickr, but it doesn't work. We need to think of something better.

Please review it in order to see if something's wrong.

k4nar commented:
Can you update the README ?

k4nar commented on drivers/flickr/manifest.json:
```diff
@@ -0,0 +1,28 @@
+	    "name": "Changes timer",
+	    "description": "The amount of time you want Onitu to wait between each check for new files on your Flickr account, in seconds. Defaults to 30.",
+	    "type": "integer",
+	    "default": 30
```
I think we should choose a saner default. Something like '120' or even '300' should be enough.

k4nar commented on drivers/flickr/onitu_flickr/flickr_driver.py:
```diff
@@ -0,0 +1,383 @@
+        self.connectAPI()
+        self.user_id = self.flickr_api.flickr.token_cache.token.user_nsid
+
+    def connectAPI(self):
```
No CamelCase in the methods names :) .

k4nar commented on drivers/flickr/onitu_flickr/flickr_driver.py:
```diff
@@ -0,0 +1,383 @@
+                warn = u"An unknown Flickr error occurred: {}".format(
+                       fe.message)
+                if fe.message == u"Error: 1: Photoset not found":
+                    warn = (u"The album {} doesn't exist on Flickr."
```
Couldn't we create it then ?

Scylardor replied:
No, because in order to create a Flickr album we MUST set a primary photo to put in this album. The check changes thread isn't really able to do that...

I thought it's a 'not-so-abnormal' situation since e.g. Flickr deletes an album when you delete all photos in it (and it could happen when doing so from Onitu). So I think if the user wants to clean up a bit, it can be a normal situation that the album's temporarily gone.

The upload_file handler manages this situation. If the album we want to upload a photo in doesn't exist (or doesn't exist _anymore_), we (re-)create it.

k4nar replied:
:+1:

Scylardor committed:
Fix problems for pull request: update README, no camelCase...

Scylardor committed:
No camelCase in check changes thread method names

Scylardor commented:
I've done the following:
- Updated the README
- No more CamelCase in method names.

k4nar commented on drivers/flickr/MANIFEST.in:
```diff
@@ -0,0 +1,3 @@
+include README
```
Same thing than for SFTP, should be README.md

Scylardor committed:
Flickr: fix crash in check changes thread

The thread handled well the fact that its album didn't exist at
runtime, but not at initialization.

Scylardor committed:
Flickr: Fix a bug in upload_file when replacing deleted photos

If we already have a photo ID on Onitu of uploaded files, we call
the flickr API call replace instead of upload. But if this file has
been deleted on Flickr, replace fails. In that case, we just use a
regular upload.

- Update MANIFEST.in
